draft-irtf-cfrg-aegis-aead.md

```diff
@@ -925,13 +938,536 @@ t = S3 ^ (LE64(ad_len_bits) || LE64(msg_len_bits))
+
+for i in 0..D:
+    V'[0,i] = AESRound(V[5,i], V[0,i] ^ m[i])
+    V'[1,i] = AESRound(V[0,i], V[1,i])
+    V'[2,i] = AESRound(V[1,i], V[2,i])
+    V'[3,i] = AESRound(V[2,i], V[3,i])
+    V'[4,i] = AESRound(V[3,i], V[4,i])
+    V'[5,i] = AESRound(V[4,i], V[5,i])
+
+    V[0,i]  = V'[0,i]
+    V[1,i]  = V'[1,i]
+    V[2,i]  = V'[2,i]
+    V[3,i]  = V'[3,i]
+    V[4,i]  = V'[4,i]
+    V[5,i]  = V'[5,i]
+~~~
+
+### The Absorb Function
+
+~~~
+Absorb(ai)
+~~~
+
+The `Absorb` function is similar to the AEGIS-256 `Absorb` function, but absorbs `R` bits instead of 128 bits.
+
+Steps:
+
+~~~
+Update(ai)
+~~~
+
+### The Enc Function
+
+~~~
+Enc(xi)
+~~~
+
+The `Enc` function is similar to the AEGIS-256 `Enc` function, but encrypts `R` bits instead of 128 bits.
+
+Steps:
+
+~~~
+z = {}
+for i in 0..D:
+    z = z || (V[1,i] ^ V[4,i] ^ V[5,i] ^ (V[2,i] & V[3,i]))
+
+Update(xi)
+
+ci = xi ^ z
+
+return ci
+~~~
+
+### The Dec Function
+
+~~~
+Dec(ci)
+~~~
+
+The `Dec` function is similar to the AEGIS-256 `Dec` function, but decrypts `R` bits instead of 128 bits.
+
+Steps:
+
+~~~
+z = {}
+for i in 0..D:
+    z = z || (V[1,i] ^ V[4,i] ^ V[5,i] ^ (V[2,i] & V[3,i]))
+
+xi = ci ^ z
+
+Update(xi)
+
+return xi
+~~~
+
+### The DecPartial Function
+
+~~~
+DecPartial(cn)
+~~~
+
+The `DecPartial` function is similar to the AEGIS-256 `DecPartial` function, but decrypts up to `R` bits instead of 128 bits.
+
+Steps:
+
+~~~
+z = {}
+for i in 0..D:
+    z = z || (V[1,i] ^ V[4,i] ^ V[5,i] ^ (V[2,i] & V[3,i]))
+
+t = ZeroPad(cn, R)
+out = t ^ z
+
+xn = Truncate(out, |cn|)
+
+v = ZeroPad(xn, 128 * D)
+Update(v)
+
+return xn
+~~~
+
+### The Finalize Function
+
+~~~
+Finalize(ad_len_bits, msg_len_bits)
+~~~
+
+The `Finalize` function finalizes every AEGIS-256 instance, and combines the resulting authentication tags using the bitwise exclusive OR operation.
+
+Steps:
+
+~~~
+t = {}
+u = LE64(ad_len_bits) || LE64(msg_len_bits)
+for i in 0..D:
+    t = t || (V[3,i] ^ u)
+
+Repeat(7, Update(t, t))
+
+if tag_length == 16: # 128 bits
+    tag = ZeroPad({}, 128)
+    for i in 0..D:
+        tag = tag ^ V[0,i] ^ V[1,i] ^ V[2,i] ^ V[3,i] ^ V[4,i] ^ V[5,i]
+
+else:                # 256 bits
+    tag0 = ZeroPad({}, 128)
+    tag1 = ZeroPad({}, 128)
+    for i in 0..D:
+        tag0 = tag0 ^ V[0,i] ^ V[1,i] ^ V[2,i]
+        tag1 = tag1 ^ V[3,i] ^ V[4,i] ^ V[5,i]
+    tag = tag0 || tag1
+
+return tag
+~~~
+
+## Implementation considerations
+
+AEGIS-128X and AEGIS-256X with a degree of `1` are indentical to AEGIS-128L and AEGIS-256X.
+
+This property can be used to reduce the code size of a generic implementation.
+
+In AEGIS-128X, `V` can be represented as eight 256-bit registers (for AEGIS-128X2), or eight 512-bit registers (for AEGIS-128X4). And in AEGIS-256X, `V` can be represented as six 256-bit registers (for AEGIS-256X2), or six 512-bit registers (for AEGIS-256X4).
+
+With this representation, loops over `0..D` in the above pseudo-code can be replaced by vector instructions.
```

Comment:
Could be put on the same line as the previous sentence assuming that's readable.

`pseudo-code` can be written as one word without the hyphen.

Reply:
Done.